### PR TITLE
Add new command eaf-open-mail

### PR DESCRIPTION
**Pull request by JulienMasson**
The goal of this command is to open any html mail in EAF browser.
Several Mail User Agent are supported in this patch:
- gnus
- mu4e
- notmuch

WARNING:
A temporary file is created to store html contents.
This file is NOT auto deleted once the EAF browser is closed.

Signed-off-by: Julien Masson <massonju.eseo@gmail.com>

### CODE REVIEW

**Comment by JulienMasson** [comment]
This patch is related to the discussion #181 

This is a PoC, please let me know if you prefer to handle this differently :)

**Comment by manateelazycat** [comment]
We can add the interface to open html string in eaf browser，then we can avoid temp html file

File: eaf.el [hunk 1]
```diff
@@ -939,6 +949,67 @@ of `eaf--buffer-app-name' inside the EAF buffer."
     (switch-to-buffer buf)
     (other-window +1)))
 
+(defun eaf--gnus-htmlp (part)
+  (when-let ((type (mm-handle-type part)))
+    (string= "text/html" (car type))))
+
+(defun eaf--nomtuch-htmlp (part)
```
Comment (MatthewZMD) [inline eaf.el]: I think you misspelled notmuch

**Comment by MatthewZMD** [comment]
Thanks for the patch! I am in the office right now and cannot properly test it, but I think I've found a small typo, please see the review comment.

File: eaf.el [hunk 2]
```diff
@@ -939,6 +949,67 @@ of `eaf--buffer-app-name' inside the EAF buffer."
...
+	      (body (plist-get msg :body))
+	      (parts (car body))
+	      (content (plist-get parts :content))
+	      (part (if (listp content)
+			(cl-find-if #'eaf--nomtuch-htmlp content)
+		      (when (eaf--nomtuch-htmlp parts)
+			parts))))
+    (notmuch-get-bodypart-text msg part notmuch-show-process-crypto)))
+
+(defun eaf-open-mail ()
```
Comment (MatthewZMD) [inline eaf.el]: Maybe name it something like `eaf-open-mail-as-html` ?

**Comment by MatthewZMD** [comment]
> This file is NOT auto deleted once the EAF browser is closed.

I think you can simply delete the temporary file after calling `eaf-open`, the EAF buffer will not get deleted.

Apparently my `mail-user-agent` defaults to `message-user-agent`, I have to manually put `(setq mail-user-agent (mu4e-user-agent))` in my config to make this work. Is it possible to make it automatically identify which MUA currently in? Lile using `(derived-mode-p 'mu4e-view-mode)`

**Comment by MatthewZMD** [comment]
> avoid temp html file

We could also do this small hack
```
(eaf-open file)
(run-at-time 1 nil (lambda () (delete-file file)))
```

**Comment by manateelazycat** [comment]
> > avoid temp html file
> 
> We can also do this small hack, it won't hurt ;)
> 
> ```
> (eaf-open file)
> (run-at-time 1 nil (lambda () (delete-file file)))
> ```

It's best delete temp file when QtWebEngine load page finish.

**Comment by MatthewZMD** [comment]
> when QtWebEngine load page finish.

Is it possible to track when does QtWebEngine finish loading?

**Comment by manateelazycat** [comment]
> > when QtWebEngine load page finish.
> 
> Is it possible to track when does QtWebEngine finish loading?

In file core/browser.py: 
```
self.buffer_widget.loadFinished.connect(self.stop_progress)
```

**Comment by manateelazycat** [comment]
There have two way:

1. Pass temp html file to eaf-open, then delete temp html file after QtWebEngine load finish.
2. Build a new interface by https://doc.qt.io/qt-5/qwebengineview.html#setHtml , then pass html string to EAF process directly.

I prefer temp html file since it's ugly that pass big data through DBus.

**Comment by JulienMasson** [comment]
Another solution is to delete the temporary file when we receive `eaf--update-buffer-title` signal ?
That's a little bit dirty since it's not really the purpose of this function ... 

Maybe another dbus signal like `eaf--buffer-created` can be used/created for this kind of use case ?

File: eaf.el [hunk 3]
```diff
@@ -939,6 +949,67 @@ of `eaf--buffer-app-name' inside the EAF buffer."
...
+In that way the corresponding function will be called to retrieve the HTML part of
+the current mail."
+  (interactive)
+  (when-let* ((get-html (eaf--get-html-func))
+	      (html (funcall get-html))
+	      (filename (concat (make-temp-name "eaf-mail-") ".html"))
+	      (file (concat (temporary-file-directory) filename)))
+    (with-temp-file file
+      (insert html))
+    (eaf-open file)))
```
Comment (manateelazycat) [inline eaf.el]: I have push new patch https://github.com/manateelazycat/emacs-application-framework/commit/8d9d78f2adf2521b18e41137cae62127b4c1219e

Change (eaf-open file) to 
```Elisp
(eaf-open file "browser" "temp_html_file")
```

Then EAF will read content of html to render, and delete temp file automatically

**Comment by manateelazycat** [comment]
> Another solution is to delete the temporary file when we receive `eaf--update-buffer-title` signal ?
> That's a little bit dirty since it's not really the purpose of this function ...
> 
> Maybe another dbus signal like `eaf--buffer-created` can be used/created for this kind of use case ?

No, it's hacking way, use 
```
(eaf-open file "browser" "temp_html_file")
```

**Comment by JulienMasson** [comment]
Thanks @manateelazycat I updated my commit with your last change.

**Comment by manateelazycat** [comment]
thanks man

**Comment by MatthewZMD** [comment]
Great, but I don't think my precious comment has been addressed yet 

> Apparently my `mail-user-agent` defaults to `message-user-agent`, I have to manually put `(setq mail-user-agent (mu4e-user-agent))` in my config to make this work.
> 
> Is it possible to make it automatically identify which MUA currently in? Like using `(derived-mode-p 'mu4e-view-mode)`

**Comment by MatthewZMD** [comment]
Also, strangely I am receiving errors that I haven't receive yesterday

```
Debugger entered--Lisp error: (void-variable eaf-mu4e-get-html)
  eval(eaf-mu4e-get-html)
  (and t (eval (assoc-default mail-user-agent eaf-mua-get-html)))
  (let* ((get-html (and t (eval (assoc-default mail-user-agent eaf-mua-get-html))))) (if get-html get-html (error "[EAF] Mail User Agent \"%s\" not supported! See all ..." mail-user-agent)))
  eaf--get-html-func()
  (and t (eaf--get-html-func))
  (let* ((html (and t (eaf--get-html-func))) (file (and html (concat (temporary-file-directory) (make-temp-name "eaf-mail-") ".html")))) (if file (progn (let ((temp-file file) (temp-buffer (generate-new-buffer " *temp file*"))) (unwind-protect (prog1 (save-current-buffer (set-buffer temp-buffer) (insert html)) (save-current-buffer (set-buffer temp-buffer) (write-region nil nil temp-file nil 0))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))) (eaf-open file "browser" "temp_html_file"))))
```

Looks like `eaf-mu4e-get-html` is been evaluated as a variable instead of a function

**Comment by MatthewZMD** [comment]
> Also, strangely I am receiving errors that I haven't receive yesterday
> 
> ```
> Debugger entered--Lisp error: (wrong-type-argument char-or-string-p eaf-mu4e-get-html)
>   insert(eaf-mu4e-get-html)
>   (save-current-buffer (set-buffer temp-buffer) (insert html))
>   (prog1 (save-current-buffer (set-buffer temp-buffer) (insert html)) (save-current-buffer (set-buffer temp-buffer) (write-region nil nil temp-file nil 0)))
> ```
> 
> Looks like `eaf-mu4e-get-html` didn't get evaluated as a function

Nevermind, this error is introduced from b6f54c8, I can fix it

**Comment by JulienMasson** [comment]
> Instead of letting the user customize mail-user-agent themselves, we should automatically detect the current MUA

@MatthewZMD yes sorry I forgot about this one.
Well it adds more complexity because you can have different `major-mode` for a single MUA.
For example in notmuch the major-mode of the buffer can be `notmuch-show-mode` or `notmuch-tree-mode`.
Same case for gnus if your current buffer is the "Article" or the "Summary", the major mode is different.
That's why I made the choice to only check `mail-user-agent`.
When you require notmuch, `mail-user-agent` is set automatically:
https://github.com/notmuch/notmuch/blob/master/emacs/notmuch.el#L1133
But not all the MUA do this ...

So as you want, we change the logic and check the major mode if you prefer

**Comment by MatthewZMD** [comment]
Sounds fair. Another possible solution to reduce complexity is to check if the major-mode start with gnus, notmuch mu4e prefixes

**Pull request merged**